1235 - Include missing bulk-only samples in metadata-only downloads

nozomione opened the pull request:
## Issue Number

Closes #1235

> [!note]
> As a fist step, we've addressed the missing samples in `metadata-only` downloads (portal-wide and project-level metadata files). 
> 
>  For `single_cell_metadata.tsv` changes, we'll handle them in a separate issue. Please see [this comment](https://github.com/AlexsLemonade/scpca-portal/issues/1235#issuecomment-2867882672) for an update  `single_cell_metadata.tsv`.  Thank you!

## Purpose/Implementation Notes

Previously, 4 bulk-only samples were missing from the generated metadata download files.

This PR ensures that all bulk samples are now included in the following `metadata-only` downloads:
- Portal-wide metadata (i.e., `PORTAL_ALL_METADATA_<accessed-date>.zip`)
- Project-level metadata (i.e., `<project-accession-id>_ALL_METADATA_<accessed-date>.zip`)

Changes made:

- Created bulk library objects in the database during metadata loading
- Assigned the modality `BULK_RNA_SEQ` to bulk libraries
- Updated `Project::get_libraries` to return all libraries for metadata-only downloads

Other change:
- Removed a `TODO` comment and a line that manually added the project ID, since the project ID is now included in metadata JSON file. This prevents the `project_id` column to from being added to the generated metadata TSV files.

For the UI ([comment](https://github.com/AlexsLemonade/scpca-portal/pull/1254#pullrequestreview-2834204379)):
- Excluded bulk libraries from the project summaries in `Project::update_project_aggregate_properties`
- **FE:** Added a check `hasSingleCellData` to render the `"Single-cell"` string in the `ProjectSamplesTable` component

## Types of changes

- Refactor (addresses code organization and design mentioned in corresponding issue)
- New feature (non-breaking change which adds functionality)

## Functional tests

The following tests have been updated:

- `test_loader`
- `test_create_portal_metadata`
- `test_dataset`
- `test/expected_values/`

## Checklist

<!-- Put an `x` in the boxes that apply. -->

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

davidsmejia reviewed (commented):
This looks good to me for now. Just want to confirm with other people that the updated summaries are not going to be an issue. Will approve or request changes based on the outcome of that discussion.

davidsmejia commented:
Ok for this PR we should remove the bulk libraries when we generate the aggregate values. This might be more involved than just filtering out bulk libraries in `Project.py:357` since we want to keep the changes in the samples table.

I think the best approach would just be to filter out bulk seq_unit in the loop at `Project.py:480`

davidsmejia commented:
Before merging this we should decide if we can address bulk only samples having `single-cell` in modalities here or open a new issue to address that since it is not a side effect of this change.

nozomione commented:
> Before merging this we should decide if we can address bulk only samples having `single-cell` in modalities here or open a new issue to address that since it is not a side effect of this change.

Per our discussion, we included the fix here since it was a small FE update (I just added a simple check), and also updated the `Project::update_project_aggregate_properties` method and the corresponding test.

This PR is now ready for your review. Thank you, David!

nozomione commented on api/scpca_portal/test/expected_values/project_SCPCP999990.py:
```diff
@@ -41,8 +41,8 @@ class Project_SCPCP999990:
-        "seq_units": "cell, spot",
-        "technologies": "10Xv3, visium",
+        "seq_units": "bulk, cell, spot",
+        "technologies": "10Xv3, paired_end, visium",
```
🗒️  Based on the team's discussion in Slack:

We'll filter out bulk libraries from the `summaries` in the Project Details but retain them in the Sample Details view. Additionally, update the FE to display the` Single-cell` string in the `Modalities` column for each row only if the sample includes single-cell data.

davidsmejia commented on client/src/components/ProjectSamplesTable.js:
```diff
@@ -172,7 +172,8 @@ export const ProjectSamplesTable = ({
       Header: 'Modalities',
-      accessor: ({ modalities }) => ['Single-cell', ...modalities].join(', ')
+      accessor: ({ has_single_cell_data: hasSingleCellData, modalities }) =>
+        [hasSingleCellData && 'Single-cell', ...modalities].join(', ')
```
Do you think we could iterate over modalities and call getReadable? You might need to make an entry for `BULK_RNA_SEQ`

nozomione replied:
It looks the values are already returned in a human-readable format from the API. We've prefixed the `"Single-cell" as a default value since it's not included in the API response.

e.g. ) https://api.scpca.alexslemonade.org/v1/samples/SCPCS000050

```js
"modalities": [
        "Bulk RNA-seq",
        "CITE-seq"
    ],
```

Alternatively, I can add the `Single Cell` to the `modalities` property in the `Sample` model so it's rendered automatically without needing to manually insert the value.
```py

    @property
    def modalities(self):
        attr_name_modality_mapping = {
            "has_single_cell_data": Modalities.SINGLE_CELL, # Add here
            "has_bulk_rna_seq": Modalities.BULK_RNA_SEQ,
            "has_cite_seq_data": Modalities.CITE_SEQ,
            "has_multiplexed_data": Modalities.MULTIPLEXED,
            "has_spatial_data": Modalities.SPATIAL,
        }

        return sorted(
            [
                modality_name.label
                for attr_name, modality_name in attr_name_modality_mapping.items()
                if getattr(self, attr_name)
            ]
        )
```

Let me know which you prefer. Thank you, David!

nozomione replied:
🗒️ We'll refactor the BE to convert these human-readable attributes into tokens and return a list of tokens to the FE. On the FE, we'll use `getReadable` to convert the tokens back into a human-readable format, which we'll handle in separate issues.

For the time being, we'll leave the implementation as it is.

davidsmejia commented on api/scpca_portal/models/project.py:
```diff
@@ -433,16 +450,28 @@ def update_project_aggregate_properties(self):
         summaries_counts = Counter()
         technologies = set()
 
+        # Get the seq_unit and technology values of bulk libraries to exclude them later
+        bulk_libraries = Library.objects.filter(modality=Modalities.BULK_RNA_SEQ)
+        bulk_seq_units = {library.metadata.get("seq_unit") for library in bulk_libraries}
+        bulk_technologies = {library.metadata.get("technology") for library in bulk_libraries}
```
```suggestion
        # Get the seq_unit and technology values of bulk libraries to exclude them later
        bulk_libraries = Library.objects.filter(modality=Modalities.BULK_RNA_SEQ)
        bulk_seq_units = set(bulk_libraries.values_list('metadata__seq_unit', flat=True))
        bulk_technologies = set(bulk_libraries.values_list('metadata__technology', flat=True))
```

This should work as well as if these were not defined for a bulk library we would have a problem.

davidsmejia commented on api/scpca_portal/models/project.py:
```diff
@@ -484,6 +513,8 @@ def update_project_aggregate_properties(self):
             project_summary.sample_count = count
             project_summary.save(update_fields=("sample_count",))
 
+        return self.summaries.all()
```
```suggestion
```
Was this just added for testing? If we don't use it we should remove.

nozomione replied:
Yes, it was added for testing - now removed 👍

davidsmejia reviewed (approved):
LGTM some minor feedback to look at before merging though.

davidsmejia commented on api/scpca_portal/models/project.py:
```diff
@@ -433,16 +450,28 @@ def update_project_aggregate_properties(self):
+            sample_seq_units = set(sample.seq_units.split(", ")) - bulk_seq_units
+            sample_technologies = set(sample.technologies.split(", ")) - bulk_technologies
+
+            if not sample_seq_units or not sample_technologies:
+                continue
```
we can remove this

avrohomgottlieb reviewed (approved):
This is an awesome PR! Everything looks great.

The one comment that I would make is that now that we're treating the bulk tsv files as input metadata files, you should add `".tsv"` to the `METADATA_EXTENTIONS` set in common.py. As well, you should check that all of the tests pass for `OriginalFile`, as this change will activate the `is_metadata=True` attribute in `OriginalFile` for bulk tsv files.  Bulk tsv files were actually classified as `is_metadata=False` (this happens in `InputBucketS3KeyInfo`) due to the `.tsv` extension not being included inside of `METADATA_EXTENSIONS`, which we only used for input metadata files.

nozomione commented:
> This is an awesome PR! Everything looks great.
> 
> The one comment that I would make is that now that we're treating the bulk tsv files as input metadata files, you should add `".tsv"` to the `METADATA_EXTENTIONS` set in common.py. As well, you should check that all of the tests pass for `OriginalFile`, as this change will activate the `is_metadata=True` attribute in `OriginalFile` for bulk tsv files. Bulk tsv files were actually classified as `is_metadata=False` (this happens in `InputBucketS3KeyInfo`) due to the `.tsv` extension not being included inside of `METADATA_EXTENSIONS`, which we only used for input metadata files.

Thank you for your insight, Avrohom! I’ve added the `.tsv` file extension to `METADATA_EXTENSIONS` as you suggested 👍 Additionally, I’ve excluded the bulk files from the `InputBucketS3KeyInfo._is_metadata` flag, as they are treated as input metadata files but still need to be included in downloads - this ensures that all tests pass.
